### PR TITLE
[WICKET-7035] fileCountMax is added

### PR DESCRIPTION
**Pull request by solomax**
POC, please review :)

I'll add some tests a bit later

### CODE REVIEW

**Review by martin-g — APPROVED** [review]
I think it would be also fine to use `long` (primitive) as https://github.com/apache/commons-fileupload/blob/c4f32a13298f6b9561c136b61b23c9e2c6d82368/src/main/java/org/apache/commons/fileupload/FileUploadBase.java#L172
I doubt the default value of `-1` will ever change.

**Comment by solomax** [comment]
@martin-g I would like to add `setFileCountMax()` to `MultiFileUploadField` (at line 225)
Shall I add logic for the case of multiple `MultiFileUploadField` in the form?

**Comment by solomax** [comment]
Other than that everything seems to work :)

**Comment by martin-g** [comment]
> I would like to add `setFileCountMax()` to `MultiFileUploadField` (at line 225)

You mean to call `form.setFileCountMax(this.max)` ?
Why ? I think this is responsibility of the developer to set the max per Form. Wicket should not try to calculate it. Or should it ?!

**Comment by solomax** [comment]
> You mean to call form.setFileCountMax(this.max) ?

yes

> Why ? I think this is responsibility of the developer to set the max per Form. Wicket should not try to calculate it. Or should it ?!

in this particular case `max` is set for `MultiFileUploadField` but *not* for the form
so now developer should pass max for both form and field, which is Ok in case there are multiple file inputs, but seems to be sort-of-code-duplication in case there is only one `MultiFileUploadField` for the form

**Comment by martin-g** [comment]
In that case maybe it would be better to not do anything at line 225 but add logic to `Form#onBeforeRender()` (?!) that does something like:
```
 # pseudo code
if (maxFileCount == -1) {
  AtomicLong accumulator = new AtomicLong(0);
  visitAllMultipartFormComponents((mpfc) -> {accumulator.add(mpfc.getMax())});
  setMaxFileCount(accumulator.get());
}
```

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/Form.java [hunk 1]
```diff
@@ -1866,6 +1904,29 @@ protected void onBeforeRender()
 		// clear multipart hint, it will be reevaluated by #isMultiPart()
 		this.multiPart &= MULTIPART_HARD;
 
+		if (!fileCountMax.isPresent())
+		{
+			AtomicLong accumulator = new AtomicLong(0);
+			visitChildren(MultiFileUploadField.class, new IVisitor<MultiFileUploadField, Void>()
```
[the remark this answers is not in the document]
Comment (solomax) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/Form.java:1910]: This is fair 
Additional case would be inline Forms with different file inputs ...

Shall I revert my last commits, so the user will take care of this? :)

[the remark this answers is not in the document]
Comment (martin-g) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/Form.java:1910]: Another option is to visit the FileUploadFields first and if it does not have `multiple` attribute then count it as `1`. It it has `multiple` then stop and do nothing more.

[the remark this answers is not in the document]
Comment (solomax) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/Form.java:1928]: I've added the code, but some tests are broken :(
I guess the better option would be to revert latest changes ....

[the remark this answers is not in the document]
Comment (martin-g) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/Form.java:1928]: OK!

**Comment by solomax** [comment]
@martin-g, @bitstorm  I guess this one should be backported to 8.x as well?

**Comment by martin-g** [comment]
+1

**Comment by solomax** [comment]
I'll merge this one into master after https://github.com/apache/wicket/pull/565

**Pull request closed without merging**